File: examples/.net/apps/portal/registry.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using AppProtocol.Example.DotNet.Cases.Tasks.TaskCreate;
using AppProtocol.Example.DotNet.Cases.Tasks.TaskList;
using AppProtocol.Example.DotNet.Cases.Tasks.TaskMove;
using AppProtocol.Example.DotNet.Core.Shared;
using AppProtocol.Example.DotNet.Packages.DesignSystem;

namespace AppProtocol.Example.DotNet.Apps.Portal;

public sealed class PortalConfig
{
    public int Port { get; init; } = 3002;
    public string ApiBaseUrl { get; init; } = "http://localhost:3000";
}

public sealed class PortalRegistry : IAppRegistry
{
    public required IDictionary<string, IDictionary<string, AppCaseSurfaces>> Cases { get; init; }
    public required IDictionary<string, object?> Providers { get; init; }
    public required IDictionary<string, object?> Packages { get; init; }
}

public static class PortalRegistryFactory
{
    public static PortalRegistry Create(PortalConfig? config = null, HttpClient? httpClient = null)
    {
        var resolvedConfig = config ?? new PortalConfig();
        var resolvedHttpClient = httpClient ?? new HttpClient
        {
            BaseAddress = new Uri(EnsureTrailingSlash(resolvedConfig.ApiBaseUrl)),
        };

        return new PortalRegistry
        {
            Cases = new Dictionary<string, IDictionary<string, AppCaseSurfaces>>(StringComparer.Ordinal)
            {
                ["tasks"] = new Dictionary<string, AppCaseSurfaces>(StringComparer.Ordinal)
                {
                    ["task_create"] = new() { Ui = typeof(TaskCreateUi) },
                    ["task_list"] = new() { Ui = typeof(TaskListUi) },
                    ["task_move"] = new() { Ui = typeof(TaskMoveUi) },
                },
            },
            Providers = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["port"] = resolvedConfig.Port,
                ["httpClient"] = new PortalHttpAdapter(resolvedHttpClient),
            },
            Packages = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["designSystem"] = new DesignSystemPackage(),
            },
        };
    }

    private static string EnsureTrailingSlash(string baseUrl)
    {
        return baseUrl.EndsWith("/", StringComparison.Ordinal) ? baseUrl : $"{baseUrl}/";
    }
}

internal sealed class PortalHttpAdapter : IAppHttpClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        WriteIndented = true,
    };

    private readonly HttpClient _httpClient;

    public PortalHttpAdapter(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task<object?> RequestAsync(AppHttpRequest request, CancellationToken cancellationToken = default)
    {
        using var message = new HttpRequestMessage(new HttpMethod(request.Method), request.Url);
        if (request.Body is not null)
        {
            message.Content = JsonContent.Create(request.Body, options: JsonOptions);
        }

        using var response = await _httpClient.SendAsync(message, cancellationToken);
        if (response.StatusCode == HttpStatusCode.NoContent)
        {
            return null;
        }

        if (response.Content.Headers.ContentLength is 0)
        {
            return null;
        }

        using var document = await JsonDocument.ParseAsync(
            await response.Content.ReadAsStreamAsync(cancellationToken),
            cancellationToken: cancellationToken);

        var payload = document.RootElement.Clone();
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                ExtractErrorMessage(payload) ??
                $"HTTP {(int)response.StatusCode} while requesting {request.Url}");
        }

        if (payload.ValueKind == JsonValueKind.Object &&
            payload.TryGetProperty("success", out var successProperty) &&
            successProperty.ValueKind is JsonValueKind.True or JsonValueKind.False)
        {
            if (!successProperty.GetBoolean())
            {
                throw new InvalidOperationException(ExtractErrorMessage(payload) ?? "Request failed");
            }

            if (payload.TryGetProperty("data", out var dataProperty))
            {
                return dataProperty.Clone();
            }

            return null;
        }

        return payload;
    }

    private static string? ExtractErrorMessage(JsonElement payload)
    {
        if (payload.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        if (payload.TryGetProperty("error", out var errorProperty) &&
            errorProperty.ValueKind == JsonValueKind.Object &&
            errorProperty.TryGetProperty("message", out var messageProperty) &&
            messageProperty.ValueKind == JsonValueKind.String)
        {
            return messageProperty.GetString();
        }

        if (payload.TryGetProperty("message", out var rootMessageProperty) &&
            rootMessageProperty.ValueKind == JsonValueKind.String)
        {
            return rootMessageProperty.GetString();
        }

        return null;
    }
}

File: examples/flutter/apps/portal/registry.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../cases/tasks/task_create/task_create.ui.case.dart';
import '../../cases/tasks/task_list/task_list.ui.case.dart';
import '../../cases/tasks/task_move/task_move.ui.case.dart';
import '../../core/shared/app_host_contracts.dart';
import '../../core/shared/app_infra_contracts.dart';
import '../../packages/design_system/index.dart';

class FetchHttpAdapter implements AppHttpClient {
  FetchHttpAdapter(this.baseUrl, [http.Client? client])
    : _client = client ?? http.Client();

  final String baseUrl;
  final http.Client _client;

  @override
  Future<dynamic> request(dynamic config) async {
    final request = config is Map
        ? Map<String, dynamic>.from(config)
        : <String, dynamic>{};
    final method = request['method']?.toString().toUpperCase() ?? 'GET';
    final url = request['url']?.toString() ?? '/';
    final body = request['body'];
    final uri = Uri.parse(baseUrl).resolve(url);

    final response = await _client.send(
      http.Request(method, uri)
        ..headers.addAll(<String, String>{'content-type': 'application/json'})
        ..body = body == null ? '' : jsonEncode(body),
    );

    final streamed = await http.Response.fromStream(response);
    final responseBody = streamed.body.trim().isEmpty
        ? null
        : jsonDecode(streamed.body) as dynamic;

    if (streamed.statusCode < 200 || streamed.statusCode >= 300) {
      String message = 'HTTP ${streamed.statusCode} while requesting $url';

      if (responseBody is Map) {
        final error = responseBody['error'];
        if (error is Map && error['message'] != null) {
          message = error['message'].toString();
        } else if (responseBody['message'] != null) {
          message = responseBody['message'].toString();
        }
      }

      throw Exception(message);
    }

    if (streamed.statusCode == 204 || responseBody == null) {
      return null;
    }

    if (responseBody is Map && responseBody['success'] is bool) {
      if (responseBody['success'] != true) {
        throw Exception(
          (responseBody['error'] is Map &&
                  responseBody['error']['message'] != null)
              ? responseBody['error']['message'].toString()
              : 'Request failed',
        );
      }

      return responseBody['data'];
    }

    return responseBody;
  }
}

class PortalConfig {
  const PortalConfig({required this.apiBaseUrl});

  final String apiBaseUrl;
}

class PortalRegistry implements AppRegistry {
  PortalRegistry({
    required AppCasesRegistry cases,
    required Map<String, dynamic> providers,
    required Map<String, dynamic> packages,
  }) : _cases = cases,
       _providers = providers,
       _packages = packages;

  final AppCasesRegistry _cases;
  final Map<String, dynamic> _providers;
  final Map<String, dynamic> _packages;

  @override
  AppCasesRegistry get cases => _cases;

  @override
  Map<String, dynamic> get providers => _providers;

  @override
  Map<String, dynamic> get packages => _packages;
}

PortalRegistry createRegistry(PortalConfig config) {
  return PortalRegistry(
    cases: <String, Map<String, AppCaseSurfaces>>{
      'tasks': <String, AppCaseSurfaces>{
        'task_create': AppCaseSurfaces(ui: (context) => TaskCreateUi(context)),
        'task_list': AppCaseSurfaces(ui: (context) => TaskListUi(context)),
        'task_move': AppCaseSurfaces(ui: (context) => TaskMoveUi(context)),
      },
    },
    providers: <String, dynamic>{
      'httpClient': FetchHttpAdapter(config.apiBaseUrl),
    },
    packages: <String, dynamic>{'designSystem': const DesignSystem()},
  );
}
